modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/ExtensionComponent.java

```diff
@@ -0,0 +1,204 @@
+    {
+        return metadataMediator.getMetadataKeys(getMetadataContext());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public MetadataResult<ComponentMetadataDescriptor> getMetadata() throws MetadataResolvingException
+    {
+        return metadataMediator.getMetadata();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public MetadataResult<ComponentMetadataDescriptor> getMetadata(MetadataKey key) throws MetadataResolvingException
+    {
+        return metadataMediator.getMetadata(getMetadataContext(), key);
+    }
+
+    @Override
+    public void setFlowConstruct(FlowConstruct flowConstruct)
+    {
+        this.flowConstruct = flowConstruct;
+    }
+
+    private MetadataContext getMetadataContext() throws MetadataResolvingException
+    {
+        //TODO MULE-9530: Improve Config retrieval for Metadata resolution
+        if (!StringUtils.isBlank(configurationProviderName) &&
+            muleContext.getRegistry().get(configurationProviderName) instanceof DynamicConfigurationProvider)
+        {
+            throw new MetadataResolvingException("Configuration used for Metadata fetch cannot be dynamic", FailureCode.INVALID_CONFIGURATION);
+        }
+
+        ConfigurationInstance<Object> configuration = getConfiguration(getInitialiserEvent(muleContext));
+        String cacheId = configuration.getName();
+
+        return new DefaultMetadataContext(configuration, connectionManager, metadataManager.getMetadataCache(cacheId));
+    }
+
+    /**
+     * @param event a {@link MuleEvent}
+     * @return a configuration instance for the current component with a given {@link MuleEvent}
+     */
+    protected ConfigurationInstance<Object> getConfiguration(MuleEvent event)
+    {
+        return getConfigurationProvider()
+                .map(provider -> provider.get(event))
+                .orElseGet(() -> {
+                    if (StringUtils.isBlank(configurationProviderName))
+                    {
+                        return extensionManager.getConfiguration(extensionModel, event);
+                    }
+                    return extensionManager.getConfiguration(configurationProviderName, event);
+                });
+    }
+
+    private Optional<ConfigurationProvider<Object>> getConfigurationProvider()
+    {
+        Optional<ConfigurationProvider<Object>> provider = StringUtils.isBlank(configurationProviderName)
+                                                           ? extensionManager.getConfigurationProvider(extensionModel)
+                                                           : extensionManager.getConfigurationProvider(configurationProviderName);
+        return provider;
+    }
+
+    /**
+     * Validates if the current component is valid for the set configuration.
+     * In case that the validation fails, the method will throw a {@link IllegalComponentException}
+     */
+    protected void validateComponentConfiguration()
```

Comment:
This code is overly complicated in comparison to the one you're replacing. It will also be very hard for it to provide a good error message.

I would rather have each subclass implementing this on its own way than trying to have this generic logic.